feat: Native datepicker

DrZoidberg09 opened the pull request:
This is an alternative suggestion to PR #19719

@barredterra rightfully suggested to directly move to native datepickers (for date/datetime-local/time).

Now everything is handled directly by the browser.

Advantages are:
- Native behavior on every device
- No external library needed anymore
- Less string transformation needed as date format is consistently very close to Frappe's
- Less validations needed as they are directly handled by the browser (e.g. correct date format)
- AM / PM now handled natively
- Stable + fast

Changes are:
- Date format is now handled by the user agent of the browser (should be anyway more convenient)
- Style is now based on the browser

What do you think of the third try to improve the datepicker? 😅

DrZoidberg09 committed:
Switch to native date picker

DrZoidberg09 committed:
Cleanup

DrZoidberg09 committed:
fixed date conversion

barredterra reviewed (commented):
Looks promising, thanks!

- For the `DateRange` control we might still need `air-datepicker` or find a workaround.
- Some more methods are no longer needed and can be removed:
    - `Date.get_df_options()`,
    - `DateTime.set_date_options()`
- `set_description` from  `DateTime` and `Time` could be lifted into the `Date` class
- read-only fields are not showing the formatted date yet

DrZoidberg09 committed:
various fixes and cleanup

DrZoidberg09 committed:
various fixes and cleanup

DrZoidberg09 commented:
@barredterra Thank you very much for the review!

I fixed everything besides for the DateRange topic. DateRange is indeed not possible with the native datepicker. 

As far as I can tell, it is only used by report filters and if I am not mistaken only by one report in ERPNext. Question is, if it might be something to be removed, if it is hardly used at all. The "workaround" is what is done by 99% of the cases (if I am not mistaken): Having two different filters for start and end date.

Otherwise, this could still stay with air-datepicker. 

Any opinions on that?

barredterra commented:
Yes, DateRange is used in report view and in list view filters. You can set a filter like "date, between, daterange". I think we cannot remove this.

One more issue is different timezones. Clicking the "Now" button in the native picker sets the time according to the browser's time zone (`Intl.DateTimeFormat().resolvedOptions().timeZone`). But our code assumes the timezone as per **User** record, then converts it to the timezone as per **System Settings**. If the timezone in **User** is different from the browser, we get the wrong time in system timezone.

DrZoidberg09 commented:
I will look into it. Meanwhile a maybe ignorant question: Is there any real world use case to set the user timezone to different timezone the user is actually in? Wouldn't it make more sense to set the user timezone automatically according to the timezone the user is actually in? This would be pretty simple to do.
Then both date format and timezone would fully automatically set by user's OS and browser. What do you think?

SaiFi0102 commented:
Good idea, but I would like to raise concerns over removing the JavaScript version of date picker.

1. Inconsistency: Not all browsers / OS have decent date pickers. Even air-datepicker is just decent, not great
2. Web View: Websites/Portals should have decent backwards compatibility. So maybe we can use the native date picker when available otherwise use air-datepicker?
3. Organization Locale: Removing date_format is too destructive. Date Formatting should be determined by the organization so that everyone communicates using the same format in my opinion

barredterra commented:
@SaiFi0102 most mobile clients (android) have far superior native pickers. Most desktop clients (chrome) also have good, IMO better, pickers. It will work with recent releases of all major browsers. IE 11 is of course out of reach, but I don't think we're still supporting it.

Regarding date format, I don't understand your concern. The goal of localization is that users can use their own language, own timezone, own date formats, etc. The system will handle the rest and everyone still uses the same underlying data.

DrZoidberg09 commented:
@surajshetty3416 @shariquerik @barredterra 
Any comments on the PR? Is there any interest in this PR at all? Otherwise we can close it for now and use it in private apps for the time being.
Then we can also open it again, if there is an interest at a later stage.

barredterra commented:
@DrZoidberg09 I'm still 100% in. Waiting for the issue reported in https://github.com/frappe/frappe/pull/19933#issuecomment-1417574060 to get resolved and happy to review again!

DrZoidberg09 commented:
Sorry, wasn't aware that you are waiting for reply on this topic. 

I think first we should clarify the question I already raised: Do we really need the option to set manually the user timezone? In which cases do you want to have a different user timezone than the one set in your system? 

I personally think it is much better to just rely on the system setting and handle everything through the browser / system. If we do this, the problem you describe does not exist.

Should we really keep the manual user timezone or use the automatic one from browser / system? Any opinions?

barredterra commented:
I agree that we should use the browser timezone.

> If we do this, the problem you describe does not exist.

We still need to change the current functionality that uses the **User** timezone to using the browser timezone. Date/time in the database is supposed to be in the timezone as per **System Settings**. This needs to be converted to/from the browser timezone. Maybe play around with multiple users in different timezones a bit and check if everything works.

DrZoidberg09 commented:
I will add it to this PR. Basically, it is not very hard to feed the datetime.utils with the browser timezone. Then every conversion system to/from user TZ should work as always (in JS). There no database entry is needed or maybe even better not to have it?!?
For python, where the user TZ should be in the database, it is a different question. Currently, I don't know where it is used and what would be the best way to set the timezone/update it. But I will figure it out.

DrZoidberg09 committed:
set user tz from users browser/system settings

DrZoidberg09 commented:
@barredterra I now added a function that the desk.js checks for the user_tz and sets in in the db if there are changes, e.g. if the user changes its location. Accordingly I set the timezone read_only in the user settings to inform the user, but not allowing him to tinker with it.

To me it looks like this should be the solution.

Please let me know your thoughts.

barredterra committed:
Merge branch 'develop' into native_datepicker

barredterra commented:
> Currently, I don't know where it is used

Looks like the only place where the user timezone is used in the backend is in ERPNext's **Service Level Agreement**. But  I don't think it's really justified there. The other code just provides it to the frontend, where we won't need it anymore.

> I now added a function that the desk.js checks for the user_tz and sets in in the db if there are changes

I have a feeling that this just makes everything more complicated. IMO, the backend should work entirely in the system timezone and the UI should convert to/from the user timezone.

DrZoidberg09 commented:
I cannot agree more. I think that would be the most consistent way. I just did it as I understood you as it should be in the database. 😄 

It is quite easy to change the datetime.js util to pull user timezone directly from the browser. Then the user TZ can be completely eliminated from the DB.

Happy to present another change in this PR soon.

DrZoidberg09 committed:
revert: set user tz from browser/system settings

DrZoidberg09 committed:
revert: set user tz from browser/system settings

DrZoidberg09 committed:
recognize tz from browser/system

DrZoidberg09 committed:
replace tz functions in other documents

barredterra commented on frappe/public/js/frappe/utils/datetime.js:
```diff
@@ -7,6 +7,7 @@ frappe.defaultDateFormat = "YYYY-MM-DD";
 frappe.defaultTimeFormat = "HH:mm:ss";
 frappe.defaultDatetimeFormat = frappe.defaultDateFormat + " " + frappe.defaultTimeFormat;
 moment.defaultFormat = frappe.defaultDateFormat;
+frappe.defaultUserTZ = Intl.DateTimeFormat().resolvedOptions().timeZone
```
A better name would be _userTimezone_, the "default" part of this has no meaning here.

barredterra commented on frappe/core/doctype/user/user.js:
```diff
@@ -328,7 +328,7 @@ frappe.ui.form.on("User", {
 	},
 	on_update: function (frm) {
-		if (frappe.boot.time_zone && frappe.boot.time_zone.user !== frm.doc.time_zone) {
+		if (frappe.boot.time_zone && frappe.defaultUserTZ !== frm.doc.time_zone) {
```
```suggestion
		if (frappe.defaultUserTZ !== frm.doc.time_zone) {
```

barredterra commented on frappe/public/js/frappe/form/controls/date.js:
```diff
@@ -155,30 +57,23 @@ frappe.ui.form.ControlDate = class ControlDate extends frappe.ui.form.ControlDat
+		super.set_description();
+	}
+	get_user_time_zone() {
+		return frappe.boot.time_zone ? frappe.defaultUserTZ : frappe.sys_defaults.time_zone;
```
```suggestion
		return frappe.defaultUserTZ || frappe.sys_defaults.time_zone;
```

barredterra commented on frappe/public/js/frappe/utils/datetime.js:
```diff
@@ -16,11 +17,11 @@ $.extend(frappe.datetime, {
 		// system time zone and then convert the string to user time zone(from User doctype).
 		let date_obj = null;
-		if (frappe.boot.time_zone && frappe.boot.time_zone.system && frappe.boot.time_zone.user) {
+		if (frappe.boot.time_zone && frappe.boot.time_zone.system && frappe.defaultUserTZ) {
```
```suggestion
		if (frappe.boot.time_zone?.system && frappe.defaultUserTZ) {
```

barredterra commented on frappe/public/js/frappe/utils/datetime.js:
```diff
@@ -35,9 +36,9 @@ $.extend(frappe.datetime, {
 		// as per the location of user.
 		let date_obj = null;
-		if (frappe.boot.time_zone && frappe.boot.time_zone.system && frappe.boot.time_zone.user) {
+		if (frappe.boot.time_zone && frappe.boot.time_zone.system && frappe.defaultUserTZ) {
```
```suggestion
		if (frappe.boot.time_zone?.system && frappe.defaultUserTZ) {
```

barredterra commented on frappe/public/js/frappe/utils/datetime.js:
```diff
@@ -48,10 +49,10 @@ $.extend(frappe.datetime, {
 
 	is_system_time_zone: function () {
-		if (frappe.boot.time_zone && frappe.boot.time_zone.system && frappe.boot.time_zone.user) {
+		if (frappe.boot.time_zone && frappe.boot.time_zone.system && frappe.defaultUserTZ) {
```
```suggestion
		if (frappe.boot.time_zone?.system && frappe.defaultUserTZ) {
```

barredterra commented on package.json:
```diff
@@ -25,7 +25,6 @@
     "@vue/component-compiler": "^4.2.4",
     "@vueuse/core": "^9.5.0",
     "ace-builds": "^1.4.8",
-    "air-datepicker": "github:frappe/air-datepicker",
```
Don't we still need this dependency for the DateRange control?

DrZoidberg09 replied:
Yes, if we want to keep this form. As mentioned, it is only used in one report. And IMHO it would be better to remove it completely, but this up to the dev-team.

DrZoidberg09 committed:
Update frappe/core/doctype/user/user.js

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>

DrZoidberg09 committed:
Update frappe/public/js/frappe/form/controls/date.js

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>

DrZoidberg09 committed:
Update frappe/public/js/frappe/utils/datetime.js

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>

DrZoidberg09 committed:
Update frappe/public/js/frappe/utils/datetime.js

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>

DrZoidberg09 committed:
Update frappe/public/js/frappe/utils/datetime.js

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>

DrZoidberg09 committed:
revert: remove air-datepicker

DrZoidberg09 committed:
revert: remove air-datepicker(files added)

DrZoidberg09 commented:
@barredterra Thank you very much for review. All you comments should be now in. I also added back air-datepicker (incl. SCSS and translations). Again: I think eventually daterange could be removed and hence the rest of air-datepicker. Might be something to discuss.

DrZoidberg09 committed:
Merge branch 'frappe:develop' into native_datepicker

DrZoidberg09 committed:
clean-up of user doctype

DrZoidberg09 commented:
I cleaned-up the User DocType from time_zone related stuff and added a function to display your own timezone read_only. Eventually, there needs to be quite some more clean-up in python functions. Quite a lot functions extract the user timzone. If I remove the docfield, frappe goes crazy... but in later stages it should be removed to achieve the goal to keep the backend free from user tz...

ankush commented:
Disclaimer: I don't look into UI much so my opinions can be ignored :woozy_face: 


I've tried this in Firefox and TBH user experience is horrible. The date picker doesn't work that well, datetime picker doesn't even have time picker. 

I am not too keen on this change but over time I agree that browser default should be available, perhaps we can make this available as an experimental option from system settings? 

Some more things that need to be checked properly:
- timezones
- datetime 
- popular browsers - Chrome, Firefox, Safari :hankey:, Edge

ankush commented:
@shariquerik take the final call please.

barredterra commented:
> I've tried this in Firefox and TBH user experience is horrible. The date picker doesn't work that well, datetime picker doesn't even have time picker.

Yeah, looks like firefox and safari didn't implement a time picker yet, at least on desktop. The big improvements of native pickers will be on mobile. Btw, I'm not convinced that our current time picker is more useful than the structured input fields offered by firefox and safari.

> I am not too keen on this change but over time I agree that browser default should be available, perhaps we can make this available as an experimental option from system settings?

Making this optional would cause a big mess, I'm afraid.

I agree that timezones need to work properly. Here, the idea was to always use the browser timezone for formatting or parsing input.

batonac commented:
+1 for this feature. The browser support table [already looks pretty good](https://caniuse.com/input-datetime), and assuming that this wouldn't roll out until v15, that give some time for the desktop browsers to "catch up".

The latest Safari desktop looks pretty good for both date and time pickers:
![image](https://user-images.githubusercontent.com/4996285/230381472-a43e69fd-461c-4982-aeb6-728f4d1a883d.png)
![image](https://user-images.githubusercontent.com/4996285/230381392-d9b98015-8e57-4e5e-98f6-edf53edafb6c.png)

Firefox desktop date picker is good:
![image](https://user-images.githubusercontent.com/4996285/230381817-512d656a-1795-4fef-b293-269cfbdb495d.png)
Firefox desktop currently hides a native time picker behind a `dom.forms.datetime.timepicker` config flag. Once activated, it looks pretty good:
![image](https://user-images.githubusercontent.com/4996285/230382071-95d975d1-1768-4d9c-b9e6-003b4824724f.png)

Of course, Chrome desktop is "on a roll":
![image](https://user-images.githubusercontent.com/4996285/230383099-f4c2ccbf-32bb-4b7b-b289-7ddb1cff4d6e.png)
![image](https://user-images.githubusercontent.com/4996285/230383009-af05d27f-2926-467f-855e-681811a8e7bc.png)

Chromium-based Edge is identical to Chrome.

shariquerik commented:
@barredterra @DrZoidberg09 
Let's not go ahead with this because of mentioned reason
1. The design of the datepicker will not match our framework design (all controls with popup has a similar design. This is the main reason I don't want this.)
2. It is still not supported properly by popular browsers. For Firefox, you still have to enable some settings which users don't know.
3. Between operator is not working and only for that if we are using the current datepicker it does not make sense.
4. While Ankush and I were testing the date format was not changing when it is changed from the system settings.

By using Native datepicker we will not have to maintain it and it will be slightly faster but it is not a good enough reason to switch. According to me, the design is the dealbreaker here.

We can fix some problems in our current datepicker like the time selector and AM/PM etc.

barredterra commented:
The design would match the enclosing browser or OS. Not sure what's more important to the user, our app or the environment they're using it it. The primary reasons for this PR were less maintenance and much superior experience on mobile.

But okay, let's postpone this and hope that browsers will catch up.

DrZoidberg09 commented:
@shariquerik 

May I ask you to have a look at this PR: #19719 ?

That could be a good compromise as an intermediate step before going towards a native datepicker.

The PR includes a switch to another datepicker library with these advantages:

- Full native look using css/scss from frappe
- no need for a fork (like now with airdatepicker), completely native use of the npm package
- Much better experience on mobile (and in general imho)
- Full support for AM/PM
- Having exactly the same behavior in terms of timezone selection and handling (even though the handling here just using system timezone and then handling of local timezone purely in the browser should be superior in the long run)

Shall we consider this? I am happy to incorporate any change, you might want to have

Thank you! 😄

shariquerik commented:
Cool if it solves all our problems 👍

DrZoidberg09 commented:
How do I get write access to my other PR? I would like to check it once again myself before asking for review from your side.

barredterra commented:
@DrZoidberg09 I reopened and unlocked the old PR
